tests/migration_test.go

```diff
@@ -932,23 +933,61 @@ var _ = Describe("[rfe_id:393][crit:high][vendor:cnv-qe@redhat.com][level:system
 		Context("with an Fedora shared NFS PVC, cloud init and service account", func() {
 			var pvName string
 			var vmi *v1.VirtualMachineInstance
-			BeforeEach(func() {
+			var nfsInitialized bool
+
+			initNFS := func(ipFamily k8sv1.IPFamily) {
 				tests.SkipNFSTestIfRunnigOnKindInfra()
 				pvName = "test-nfs" + rand.String(48)
 				// Prepare a NFS backed PV
 				By("Starting an NFS POD")
 				os := string(cd.ContainerDiskFedora)
-				nfsIP := tests.CreateNFSTargetPOD(os)
+				nfsPod := tests.CreateNFSTargetPOD(os)
+
 				// create a new PV and PVC (PVs can't be reused)
 				By("create a new NFS PV and PVC")
+				nfsIP := libnet.GetPodIpByFamily(nfsPod, ipFamily)
+				ExpectWithOffset(1, nfsIP).NotTo(BeEmpty())
 				tests.CreateNFSPvAndPvc(pvName, "5Gi", nfsIP, os)
+				nfsInitialized = true
+			}
+
+			BeforeEach(func() {
+				nfsInitialized = false
 			})
 
 			AfterEach(func() {
-				// PVs can't be reused
-				tests.DeletePvAndPvc(pvName)
+				if nfsInitialized {
+					By("Deleting the VMI")
+					Expect(virtClient.VirtualMachineInstance(vmi.Namespace).Delete(vmi.Name, &metav1.DeleteOptions{})).To(Succeed())
+					By("Waiting for VMI to disappear")
+					tests.WaitForVirtualMachineToDisappearWithTimeout(vmi, 120)
+				}
+			})
```

Comment:
I find this a bit strange - e.g. why do you only remove the VMI if the nfs was instantiated ? 

I would expect the "do I remove the VM" condition to be more explicit (related to VMs) or a simple 
```golang
Expect(vmi).NotTo(BeNil(), "if there's no VMI present in the cleanup stage, its because the test exploded and we don't care.")
```

Bottom-line, while it is correct, I think you can (and should) make it more explicit.

Reply:
The issue here is that all the other non-nfs related tests have their own cleanup/don't need a clean up.
We have here a huge `Describe` with a lot of `Context`s and tests. The nfs related tests are spread over different contexts.
I will copy the `AfterEach` nfs cleanup to each one of the context containing nfs related tests, I hope it will make the clean code clearer and more explicit.

Reply:
the way the CDI tests (at least) work is that they wouldn't need this kind of cleanup, because the resources are created in a temporary namespace that is destroyed after every test, so the cleanup of the vmi can happen for free.

Reply:
Are you talking about the `BeforeTestCleanup` that happens before the next test?
We prefer the test will clean the resources it created and won't leave garbage for the next test to clean.
Anyway, afaik, the namespace cleanup is not a full destruction of the namespace. For example the nfspod won't be removed.
Also, the cleanup is not waiting for the resources to actually be removed. So in the case of the nfsPod which has the same name and uses the same HostPath between the tests, it may cause the test to be flaky.